src/alda/lisp/model/pitch.clj

```diff
@@ -17,17 +17,38 @@
   [note]
   (* 440.0 (Math/pow 2.0 (/ (- note 69.0) 12.0))))
 
+(defn- check-key
+  "Modifies the accidentals on notes to fit the key signature.
+   If there are no accidentals and this letter is in the signature, 
+   return the letter's signature accidentals, otherwise return 
+   existing accidentals"
+  [signature letter accidentals]
+   (if (empty? accidentals)
+       (get signature letter nil) 
+       (identity accidentals)))
+
 (defn pitch
   "Returns a fn that will calculate the frequency in Hz, within the context
    of the octave that an instrument is in."
-  [letter & accidentals]
-  (fn [octave & {:keys [midi]}]
+  [parsed-pitch]
+  (let [letter (first parsed-pitch) accidentals (rest parsed-pitch)]
+  (fn [octave key-signature & {:keys [midi]}]
     (let [midi-note (reduce (fn [number accidental]
+                              (log/debug (format "number %s accidental %s" number accidental))
                               (case accidental
-                                :flat  (dec number)
-                                :sharp (inc number)))
+                                :flat    (dec number)
+                                :sharp   (inc number)
+                                :natural (identity number)))
                             (midi-note letter octave)
-                            accidentals)]
+                            (check-key key-signature letter accidentals) )]
       (if midi
         midi-note
-        (midi->hz midi-note)))))
+        (midi->hz midi-note))))))
+
+(defn parse-pitch 
+  "Converts the tokenized representation of a pitch into a keyword sequence.
+   For example, ('c' '+' '+') will become (:c :sharp :sharp)"
+  [letter & accidentals]
+  (list* (keyword (str letter))
+        (map {\+ :sharp, \= :natural, \- :flat} accidentals))
+)
```

Comment:
I think doing more on the parsing side is generally a good thing -- so far I've been trying to leverage Instaparse's powerful [transformations](https://github.com/Engelberg/instaparse#transforming-the-tree) as much as possible, to help keep the rest of the code cleaner and easier to understand. The idea is that the parser picks up on a pattern of text as a particular type of node, then transforms that type of node into the type of data that we actually want to work with.

It gets a little more interesting with Alda, because the type of data that we're generating via Instaparse's transforms is runnable Clojure code, but we also have the opportunity to transform the arguments into any format we want. For instance, here, we could absolutely have the parser transform the key signature elements into a single map, or any other format we'd like.

A good REPL syntax might be something like:

    (key-signature {:e [:flat :flat] :c [:sharp]})

(which is equivalent to your idea -- I just used vectors instead of lists)

I like the idea of a hash-map as a single object that represents a key signature. Very easy to reason about. Conceptually, it makes sense, since you can't have multiple values for the same key, i.e., you can't have more than one list of accidentals for any given letter.

This is all stuff that I should be documenting in our `CONTRIBUTING.md`, but explaining it here is good practice, at least :P